core/src/main/scala/kafka/log/Log.scala

```diff
@@ -79,6 +78,17 @@ case class LogAppendInfo(var firstOffset: Long,
  *
  * @param dir The directory in which log segments are created.
  * @param config The log configuration settings
+ * @param logStartOffset The smallest offset is allowed to be exposed to kafka client.
+ *                       The logStartOffset can be updated by :
+ *                       - user's PurgeRequest
+ *                       - broker's log retention
+ *                       - broker's log truncation
```

Comment:
I include this scenario because I couldn't provide that log truncation will always be above the logStartOffset. For example, we have had bug such that log may be truncated below high watermark in case of double failure.